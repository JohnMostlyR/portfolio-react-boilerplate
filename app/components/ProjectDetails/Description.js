import styled from 'styled-components';
import { BASE_LINE_HEIGHT } from '../../styles/typography';

const Description = styled.div`
  margin: ${BASE_LINE_HEIGHT}em 0;
  max-width: 51em;
  line-height: ${BASE_LINE_HEIGHT};
  word-break: break-word;

  @media (min-width: 950px) {
    line-height: ${BASE_LINE_HEIGHT * 1.1};
  }
`;

export default Description;
