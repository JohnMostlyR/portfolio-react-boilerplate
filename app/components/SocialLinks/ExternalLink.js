import React from 'react';
import PropTypes from 'prop-types';

import StyledA from './StyledA';
import Description from './Description';

import IconCodepen from '../../components/IconCodepen';
import IconFreeCodeCamp from '../../components/IconFreeCodeCamp';
import IconGithub from '../../components/IconGithub';
import IconLinkedin from '../../components/IconLinkedin';
import IconExternalLink from '../../components/IconExternalLink';

export const VALID_ICONS = [
  'free-code-camp',
  'github',
  'linkedin',
  'linkedin-square',
  'codepen',
];

function ExternalLink({ color, description, href, faIcon }) {
  let renderThis = <span />;

  switch (faIcon) {
    case 'free-code-camp':
      renderThis = <IconFreeCodeCamp />;
      break;
    case 'github':
      renderThis = <IconGithub />;
      break;
    case 'linkedin':
    case 'linkedin-square':
      renderThis = <IconLinkedin />;
      break;
    case 'codepen':
      renderThis = <IconCodepen />;
      break;
    default:
      renderThis = <IconExternalLink />;
  }

  return (
    <StyledA
      href={href}
      target="_blank"
      rel="noopener noreferrer"
      color={color}
    >
      {renderThis}
      <Description>{description}</Description>
    </StyledA>
  );
}

ExternalLink.propTypes = {
  color: PropTypes.string,
  description: PropTypes.string,
  href: PropTypes.string,
  faIcon: PropTypes.oneOf(VALID_ICONS),
};

ExternalLink.defaultProps = {
  color: 'inherit',
  description: 'ExternalLink to',
  href: '#',
};

export default ExternalLink;
