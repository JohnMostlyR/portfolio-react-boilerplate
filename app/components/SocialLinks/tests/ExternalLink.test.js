import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ExternalLink, { VALID_ICONS } from '../ExternalLink';
import Description from '../Description';

describe('<ExternalLink>', () => {
  it('should render and match the snapshot', () => {
    const wrapper = shallow(<ExternalLink />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should adopt the "href" property', () => {
    const EXTERNAL_LINK = 'https://some-site.com';
    const wrapper = shallow(<ExternalLink href={EXTERNAL_LINK} />);
    expect(wrapper.prop('href')).toBe(EXTERNAL_LINK);
  });

  it('should adopt the "description" property', () => {
    const VALUE = 'Follow me';
    const wrapper = shallow(<ExternalLink description={VALUE} />);
    expect(
      wrapper
        .find(Description)
        .childAt(0)
        .text(),
    ).toBe(VALUE);
  });

  it('should adopt the "color" property', () => {
    const COLOR = '#bada55';
    const wrapper = shallow(<ExternalLink color={COLOR} />);
    expect(wrapper.prop('color')).toBe(COLOR);
  });

  describe('with a given "faIcon" property', () => {
    VALID_ICONS.forEach(icon => {
      it(`should render with a "${icon}" icon`, () => {
        const wrapper = shallow(<ExternalLink faIcon={icon} />);
        expect(toJson(wrapper)).toMatchSnapshot();
      });
    });
  });
});
