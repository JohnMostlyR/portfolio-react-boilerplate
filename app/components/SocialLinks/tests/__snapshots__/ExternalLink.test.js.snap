// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExternalLink> should render and match the snapshot 1`] = `
<StyledA
  color="inherit"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  <IconExternalLink />
  <Description>
    ExternalLink to
  </Description>
</StyledA>
`;

exports[`<ExternalLink> with a given "faIcon" property should render with a "codepen" icon 1`] = `
<StyledA
  color="inherit"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  <IconCodepen />
  <Description>
    ExternalLink to
  </Description>
</StyledA>
`;

exports[`<ExternalLink> with a given "faIcon" property should render with a "free-code-camp" icon 1`] = `
<StyledA
  color="inherit"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  <IconFreeCodeCamp />
  <Description>
    ExternalLink to
  </Description>
</StyledA>
`;

exports[`<ExternalLink> with a given "faIcon" property should render with a "github" icon 1`] = `
<StyledA
  color="inherit"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  <IconGithub />
  <Description>
    ExternalLink to
  </Description>
</StyledA>
`;

exports[`<ExternalLink> with a given "faIcon" property should render with a "linkedin" icon 1`] = `
<StyledA
  color="inherit"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  <IconLinkedin />
  <Description>
    ExternalLink to
  </Description>
</StyledA>
`;

exports[`<ExternalLink> with a given "faIcon" property should render with a "linkedin-square" icon 1`] = `
<StyledA
  color="inherit"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
>
  <IconLinkedin />
  <Description>
    ExternalLink to
  </Description>
</StyledA>
`;
